### PR TITLE
Updates for TLSv1.3 draft 26

### PR DESCRIPTION
**Pull request by mattcaswell**

### CODE REVIEW

**Comment by kaduk** [comment]
travis thinks there are memory leaks, at least (I didn't look at all failures)

**Comment by mattcaswell** [comment]
Fixup commit pushed addressing the travis/appveyor failures. @richsalz please reconfirm?

**Comment by richsalz** [comment]
Reconfirmed once the CI's pass.

File: ssl/record/ssl3_record_tls13.c [hunk 1]
```diff
@@ -143,7 +144,28 @@ int tls13_enc(SSL *s, SSL3_RECORD *recs, size_t n_recs, int sending)
...
+        return -1;
+    }
+
+    /* Set up the AAD */
+    if (!WPACKET_init_static_len(&wpkt, recheader, sizeof(recheader), 0)
+            || !WPACKET_put_bytes_u8(&wpkt, rec->type)
+            || !WPACKET_put_bytes_u16(&wpkt, rec->rec_version)
+            || !WPACKET_put_bytes_u16(&wpkt, rec->length + taglen)
+            || !WPACKET_finish(&wpkt)) {
+        return -1;
```
Comment (kaduk) [inline ssl/record/ssl3_record_tls13.c]: This seems to implicitly rely on knowledge of the structure of the record header/SSL3_RT_HEADER_LENGTH -- do we want to check that there's no space left in the WPACKET and/or that what we've written matches SSL3_RT_HEADER_LENGTH.  This seems relevant given that we add the AAD based on sizeof(recheader) and not the WPACKET's idea of things.

File: ssl/statem/extensions_clnt.c [hunk 2]
```diff
@@ -1780,20 +1780,20 @@ int tls_parse_stoc_supported_versions(SSL *s, PACKET *pkt, unsigned int context,
...
-         * TLSv1.3, therefore we shouldn't be getting an HRR for anything else.
-         */
-        if (version != TLS1_3_VERSION) {
-            SSLfatal(s, SSL_AD_PROTOCOL_VERSION,
-                     SSL_F_TLS_PARSE_STOC_SUPPORTED_VERSIONS,
-                     SSL_R_BAD_HRR_VERSION);
-            return 0;
-        }
+    if (context == SSL_EXT_TLS1_3_HELLO_RETRY_REQUEST)
         return 1;
```
Comment (kaduk) [inline ssl/statem/extensions_clnt.c]: side note: aren't we supposed to check that the server is consistent between HRR and SH?

File: ssl/record/ssl3_record_tls13.c (hunk 1 above)
Comment (mattcaswell) [reply ssl/record/ssl3_record_tls13.c]: Good point. I pushed a fixup commit to address this. Also realised we need to call `WPACKET_cleanup()` in the error case so I added that too.

File: ssl/statem/extensions_clnt.c (hunk 2 above)
Comment (mattcaswell) [reply ssl/statem/extensions_clnt.c]: If we've got an HRR we check that the extension contains TLS1_3_VERSION. Anything else is invalid. Later, when we get the ServerHello back after ClientHello2, we check that we negotiate TLSv1.3 (anything else is invalid after an HRR). In order for that to happen the supported_versions extension in the HRR and the ServerHello must be consistent. So, although there is no explicit consistency check, it happens implicitly.

File: test/recordlentest.c [hunk 3]
```diff
@@ -157,11 +157,7 @@ static int test_record_overflow(int idx)
         overf_expected = 0;
     }
 
-    if (idx == TEST_ENCRYPTED_OVERFLOW_TLS1_3_OK
-            || idx == TEST_ENCRYPTED_OVERFLOW_TLS1_3_NOT_OK)
-        recversion = TLS1_VERSION;
-    else
-        recversion = TLS1_2_VERSION;
+    recversion = TLS1_2_VERSION;
```
Comment (kaduk) [inline test/recordlentest.c]: I'm slightly surprised that this works with no-tls12 builds, but any issue there is not new to this version, so it's clearly okay.

File: test/recipes/70-test_sslrecords.t [hunk 4]
```diff
@@ -464,7 +464,7 @@ sub change_version
     my $proxy = shift;
 
     # We'll change a version after the initial version neg has taken place
-    if ($proxy->flight != 2) {
+    if ($proxy->flight != 1) {
```
Comment (kaduk) [inline test/recipes/70-test_sslrecords.t]: It's not immediately obvious to me why this change is needed, but it seems harmless enough...

File: test/recordlentest.c (hunk 3 above)
Comment (mattcaswell) [reply test/recordlentest.c]: It never gets this far in a no-tls12 build due to the check at the beginning of this function.

File: test/recipes/70-test_sslrecords.t (hunk 4 above)
Comment (mattcaswell) [reply test/recipes/70-test_sslrecords.t]: I hadn't updated this test at all in the original version of my PR. Test 13 *should* have failed when I made the record version change (previously we ignored the record version, now we don't). - but it didn't. At least it didn't on Linux, but it did on AppVeyor (which is how I found this). The problem was we were making the record version change too late. Due to race conditions etc (platform dependent) sometimes it failed, sometimes it didn't. Making this change caused the test to fail reliably on all platforms.

**Comment by mattcaswell** [comment]
Pushed. Thanks!
